File: electron/appUpdater.ts
import { app, shell } from 'electron';
import { autoUpdater } from 'electron-updater';
import { existsSync, readdirSync } from 'node:fs';
import path from 'node:path';

import { APP_UPDATE_CHANNELS, type AppUpdateStatus } from '../src/shared';

const isDev = !!process.env.VITE_DEV_SERVER_URL;
const GITHUB_REPO = 'scdemanett/redump-dat-filter';
const GITHUB_RELEASES_URL = `https://github.com/${GITHUB_REPO}/releases/latest`;
const UPDATE_CHECK_TIMEOUT_MS = 15_000;
const AUTO_UPDATE_TIMEOUT_MS = 30_000;

type StatusSender = (status: AppUpdateStatus) => void;

let sendStatus: StatusSender = () => {};
let lastStatus: AppUpdateStatus = { state: 'idle' };
let autoUpdaterConfigured = false;

function setStatus(status: AppUpdateStatus): void {
  lastStatus = status;
  sendStatus(status);
}

function hasWindowsInstallerUninstaller(): boolean {
  try {
    const installDir = path.dirname(process.execPath);
    return readdirSync(installDir).some((file) => /^Uninstall .+\.exe$/i.test(file));
  } catch {
    return false;
  }
}

function supportsAutoInstall(): boolean {
  if (!app.isPackaged || isDev || process.env.PORTABLE_EXECUTABLE_DIR) {
    return false;
  }

  if (!existsSync(path.join(process.resourcesPath, 'app-update.yml'))) {
    return false;
  }

  if (process.platform === 'win32') {
    return hasWindowsInstallerUninstaller();
  }

  if (process.platform === 'linux') {
    return Boolean(process.env.APPIMAGE);
  }

  if (process.platform === 'darwin') {
    return process.execPath.includes(`${path.sep}Applications${path.sep}`);
  }

  return false;
}

function withTimeout<T>(promise: Promise<T>, timeoutMs: number, message: string): Promise<T> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(message));
    }, timeoutMs);

    promise
      .then((value) => {
        clearTimeout(timer);
        resolve(value);
      })
      .catch((error: unknown) => {
        clearTimeout(timer);
        reject(error);
      });
  });
}

function isNewerVersion(latest: string, current: string): boolean {
  const parse = (value: string) =>
    value
      .replace(/^v/i, '')
      .split('.')
      .map((part) => Number.parseInt(part, 10) || 0);

  const latestParts = parse(latest);
  const currentParts = parse(current);
  const length = Math.max(latestParts.length, currentParts.length);

  for (let index = 0; index < length; index += 1) {
    const diff = (latestParts[index] ?? 0) - (currentParts[index] ?? 0);
    if (diff !== 0) {
      return diff > 0;
    }
  }

  return false;
}

function formatReleaseNotes(notes: unknown): string | undefined {
  if (typeof notes === 'string') {
    return notes.trim() || undefined;
  }

  if (Array.isArray(notes)) {
    const text = notes
      .map((entry) => {
        if (typeof entry === 'string') {
          return entry;
        }
        if (entry && typeof entry === 'object' && 'note' in entry && typeof entry.note === 'string') {
          return entry.note;
        }
        return '';
      })
      .filter(Boolean)
      .join('\n\n')
      .trim();

    return text || undefined;
  }

  return undefined;
}

function configureAutoUpdater(): void {
  if (autoUpdaterConfigured) {
    return;
  }

  autoUpdater.autoDownload = false;
  autoUpdater.autoInstallOnAppQuit = true;

  autoUpdater.on('download-progress', (progress) => {
    setStatus({
      state: 'downloading',
      percent: progress.percent
    });
  });

  autoUpdater.on('update-downloaded', (info) => {
    setStatus({
      state: 'downloaded',
      latestVersion: info.version,
      autoInstallSupported: true
    });
  });

  autoUpdater.on('error', (error) => {
    if (lastStatus.state === 'downloading') {
      setStatus({
        state: 'error',
        message: error.message
      });
    }
  });

  autoUpdaterConfigured = true;
}

async function checkViaGitHubApi(): Promise<AppUpdateStatus> {
  const currentVersion = app.getVersion();
  const autoInstallSupported = supportsAutoInstall();

  try {
    const response = await withTimeout(
      fetch(`https://api.github.com/repos/${GITHUB_REPO}/releases/latest`, {
        headers: {
          Accept: 'application/vnd.github+json',
          'User-Agent': `${app.getName()}/${currentVersion}`
        },
        signal: AbortSignal.timeout(UPDATE_CHECK_TIMEOUT_MS)
      }),
      UPDATE_CHECK_TIMEOUT_MS,
      'Timed out while checking for updates.'
    );

    if (!response.ok) {
      throw new Error(`GitHub API responded with ${response.status}`);
    }

    const payload = (await response.json()) as {
      tag_name?: string;
      html_url?: string;
      body?: string;
    };

    const latestVersion = payload.tag_name?.replace(/^v/i, '') ?? '';
    if (!latestVersion) {
      throw new Error('Latest release is missing a version tag.');
    }

    if (isNewerVersion(latestVersion, currentVersion)) {
      return {
        state: 'available',
        currentVersion,
        latestVersion,
        releaseNotes: payload.body?.trim() || undefined,
        releaseUrl: payload.html_url ?? GITHUB_RELEASES_URL,
        autoInstallSupported
      };
    }

    return {
      state: 'unavailable',
      currentVersion
    };
  } catch (error) {
    return {
      state: 'error',
      message: error instanceof Error ? error.message : 'Failed to check for updates.'
    };
  }
}

export function getAppVersion(): string {
  return app.getVersion();
}

export function getLastAppUpdateStatus(): AppUpdateStatus {
  return lastStatus;
}

export function registerAppUpdateStatusSender(sender: StatusSender): void {
  sendStatus = sender;
  sendStatus(lastStatus);
}

export function initAppUpdater(): void {
  if (!app.isPackaged || isDev) {
    setStatus({
      state: 'disabled',
      reason: 'Updates are checked in packaged builds only.'
    });
    return;
  }

  if (supportsAutoInstall()) {
    configureAutoUpdater();
  }

  setStatus({
    state: 'idle',
    currentVersion: app.getVersion()
  });

  setTimeout(() => {
    void checkForAppUpdates(false);
  }, 5000);
}

export async function checkForAppUpdates(manual: boolean): Promise<AppUpdateStatus> {
  const currentVersion = app.getVersion();

  if (!app.isPackaged || isDev) {
    const status: AppUpdateStatus = {
      state: 'disabled',
      reason: 'Updates are checked in packaged builds only.'
    };
    if (manual) {
      setStatus(status);
    }
    return status;
  }

  if (manual) {
    setStatus({ state: 'checking' });
  }

  const status = await checkViaGitHubApi();

  if (status.state === 'unavailable') {
    const resolved: AppUpdateStatus = {
      state: 'unavailable',
      currentVersion
    };
    if (manual) {
      setStatus(resolved);
    }
    return resolved;
  }

  if (manual || status.state === 'available') {
    setStatus(status);
  }

  return status;
}

async function downloadViaAutoUpdater(): Promise<AppUpdateStatus> {
  configureAutoUpdater();

  try {
    setStatus({ state: 'downloading', percent: 0 });

    const checkResult = await withTimeout(
      autoUpdater.checkForUpdates(),
      AUTO_UPDATE_TIMEOUT_MS,
      'Timed out while preparing the in-app update download.'
    );

    if (!checkResult?.updateInfo) {
      throw new Error('No in-app update is available for this install.');
    }

    await withTimeout(
      autoUpdater.downloadUpdate(),
      AUTO_UPDATE_TIMEOUT_MS,
      'Timed out while downloading the update.'
    );

    return lastStatus;
  } catch (error) {
    const status: AppUpdateStatus = {
      state: 'error',
      message: error instanceof Error ? error.message : 'Failed to download update.'
    };
    setStatus(status);
    return status;
  }
}

export async function downloadAppUpdate(): Promise<AppUpdateStatus> {
  if (lastStatus.state !== 'available') {
    return lastStatus;
  }

  if (supportsAutoInstall() && lastStatus.autoInstallSupported) {
    return downloadViaAutoUpdater();
  }

  const releaseUrl = lastStatus.releaseUrl ?? GITHUB_RELEASES_URL;
  await shell.openExternal(releaseUrl);
  return lastStatus;
}

export function installAppUpdate(): AppUpdateStatus {
  if (!supportsAutoInstall() || lastStatus.state !== 'downloaded') {
    return lastStatus;
  }

  autoUpdater.quitAndInstall();
  return lastStatus;
}

export { APP_UPDATE_CHANNELS };

File: src-tauri/src/app_updater.rs
use crate::types::AppUpdateStatus;
use serde::Deserialize;
use std::sync::Mutex;
use tauri::{AppHandle, Emitter, Manager};
use tauri_plugin_opener::OpenerExt;
use tauri_plugin_updater::UpdaterExt;

const GITHUB_REPO: &str = "scdemanett/redump-dat-filter";
const GITHUB_RELEASES_URL: &str = "https://github.com/scdemanett/redump-dat-filter/releases/latest";
const UPDATE_CHECK_TIMEOUT_MS: u64 = 15_000;

pub struct UpdaterState {
  pub last_status: Mutex<AppUpdateStatus>,
}

impl Default for UpdaterState {
  fn default() -> Self {
    Self {
      last_status: Mutex::new(AppUpdateStatus::Idle {
        current_version: None,
      }),
    }
  }
}

fn set_status(app: &AppHandle, state: &UpdaterState, status: AppUpdateStatus) {
  if let Ok(mut guard) = state.last_status.lock() {
    *guard = status.clone();
  }
  let _ = app.emit("app:update-status", status);
}

pub fn get_last_status(state: &UpdaterState) -> AppUpdateStatus {
  state
    .last_status
    .lock()
    .map(|g| g.clone())
    .unwrap_or(AppUpdateStatus::Idle {
      current_version: None,
    })
}

fn is_newer_version(latest: &str, current: &str) -> bool {
  let parse = |value: &str| -> Vec<u32> {
    value
      .trim_start_matches(|c: char| c == 'v' || c == 'V')
      .split('.')
      .map(|part| part.parse::<u32>().unwrap_or(0))
      .collect()
  };
  let latest_parts = parse(latest);
  let current_parts = parse(current);
  let len = latest_parts.len().max(current_parts.len());
  for i in 0..len {
    let l = latest_parts.get(i).copied().unwrap_or(0);
    let c = current_parts.get(i).copied().unwrap_or(0);
    if l != c {
      return l > c;
    }
  }
  false
}

fn supports_auto_install(_app: &AppHandle) -> bool {
  // Dev builds never auto-install. Packaged installers can; portable zips open GitHub Releases.
  if cfg!(debug_assertions) {
    return false;
  }

  let Ok(exe) = std::env::current_exe() else {
    return false;
  };
  let Some(dir) = exe.parent() else {
    return false;
  };

  #[cfg(target_os = "windows")]
  {
    // Tauri/Electron NSIS installs ship an uninstaller next to the binary.
    // Portable zips are just the exe, so they fall back to the release page.
    match std::fs::read_dir(dir) {
      Ok(entries) => entries.flatten().any(|entry| {
        let name = entry.file_name().to_string_lossy().to_ascii_lowercase();
        name == "uninstall.exe" || (name.starts_with("uninstall ") && name.ends_with(".exe"))
      }),
      Err(_) => false,
    }
  }

  #[cfg(target_os = "linux")]
  {
    // In-app install is supported for AppImage launches.
    std::env::var_os("APPIMAGE").is_some()
  }

  #[cfg(target_os = "macos")]
  {
    exe
      .to_string_lossy()
      .contains(&format!("{}Applications{}", std::path::MAIN_SEPARATOR, std::path::MAIN_SEPARATOR))
  }

  #[cfg(not(any(target_os = "windows", target_os = "linux", target_os = "macos")))]
  {
    let _ = dir;
    false
  }
}

#[derive(Debug, Deserialize)]
struct GitHubRelease {
  tag_name: Option<String>,
  html_url: Option<String>,
  body: Option<String>,
}

async fn check_via_github_api(app: &AppHandle) -> AppUpdateStatus {
  let current_version = app.package_info().version.to_string();
  let auto_install_supported = supports_auto_install(app);

  let client = match reqwest::Client::builder()
    .user_agent(format!("RedumpDATFilter/{current_version}"))
    .timeout(std::time::Duration::from_millis(UPDATE_CHECK_TIMEOUT_MS))
    .build()
  {
    Ok(c) => c,
    Err(e) => {
      return AppUpdateStatus::Error {
        message: e.to_string(),
      }
    }
  };

  let url = format!("https://api.github.com/repos/{GITHUB_REPO}/releases/latest");
  let response = match client
    .get(&url)
    .header("Accept", "application/vnd.github+json")
    .send()
    .await
  {
    Ok(r) => r,
    Err(e) => {
      return AppUpdateStatus::Error {
        message: e.to_string(),
      }
    }
  };

  if !response.status().is_success() {
    return AppUpdateStatus::Error {
      message: format!("GitHub API responded with {}", response.status()),
    };
  }

  let payload: GitHubRelease = match response.json().await {
    Ok(p) => p,
    Err(e) => {
      return AppUpdateStatus::Error {
        message: e.to_string(),
      }
    }
  };

  let latest_version = payload
    .tag_name
    .unwrap_or_default()
    .trim_start_matches(|c: char| c == 'v' || c == 'V')
    .to_string();

  if latest_version.is_empty() {
    return AppUpdateStatus::Error {
      message: "Latest release is missing a version tag.".into(),
    };
  }

  if is_newer_version(&latest_version, &current_version) {
    AppUpdateStatus::Available {
      current_version,
      latest_version,
      release_notes: payload
        .body
        .map(|b| b.trim().to_string())
        .filter(|s| !s.is_empty()),
      release_url: Some(
        payload
          .html_url
          .unwrap_or_else(|| GITHUB_RELEASES_URL.to_string()),
      ),
      auto_install_supported,
    }
  } else {
    AppUpdateStatus::Unavailable { current_version }
  }
}

pub fn init_app_updater(app: AppHandle) {
  let state = app.state::<UpdaterState>();
  if cfg!(debug_assertions) {
    set_status(
      &app,
      &state,
      AppUpdateStatus::Disabled {
        reason: "Updates are checked in packaged builds only.".into(),
      },
    );
    return;
  }

  set_status(
    &app,
    &state,
    AppUpdateStatus::Idle {
      current_version: Some(app.package_info().version.to_string()),
    },
  );

  let app_clone = app.clone();
  tauri::async_runtime::spawn(async move {
    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
    let state = app_clone.state::<UpdaterState>();
    let _ = check_for_app_updates(&app_clone, &state, false).await;
  });
}

pub async fn check_for_app_updates(
  app: &AppHandle,
  state: &UpdaterState,
  manual: bool,
) -> AppUpdateStatus {
  let current_version = app.package_info().version.to_string();

  if cfg!(debug_assertions) {
    let status = AppUpdateStatus::Disabled {
      reason: "Updates are checked in packaged builds only.".into(),
    };
    if manual {
      set_status(app, state, status.clone());
    }
    return status;
  }

  if manual {
    set_status(app, state, AppUpdateStatus::Checking);
  }

  let status = check_via_github_api(app).await;

  match &status {
    AppUpdateStatus::Unavailable { .. } => {
      let resolved = AppUpdateStatus::Unavailable { current_version };
      if manual {
        set_status(app, state, resolved.clone());
      }
      resolved
    }
    AppUpdateStatus::Available { .. } => {
      set_status(app, state, status.clone());
      status
    }
    other => {
      if manual {
        set_status(app, state, other.clone());
      }
      other.clone()
    }
  }
}

pub async fn download_app_update(app: &AppHandle, state: &UpdaterState) -> AppUpdateStatus {
  let last = get_last_status(state);
  let AppUpdateStatus::Available {
    release_url,
    auto_install_supported,
    latest_version,
    ..
  } = last.clone()
  else {
    return last;
  };

  if !auto_install_supported {
    let url = release_url.unwrap_or_else(|| GITHUB_RELEASES_URL.to_string());
    let _ = app.opener().open_url(url, None::<&str>);
    return get_last_status(state);
  }

  set_status(app, state, AppUpdateStatus::Downloading { percent: 0.0 });

  match app.updater() {
    Ok(updater) => match updater.check().await {
      Ok(Some(update)) => {
        let mut downloaded = 0u64;
        let result = update
          .download_and_install(
            |chunk_len, content_len| {
              downloaded += chunk_len as u64;
              if let Some(total) = content_len {
                if total > 0 {
                  let percent = (downloaded as f64 / total as f64) * 100.0;
                  set_status(app, state, AppUpdateStatus::Downloading { percent });
                }
              }
            },
            || {},
          )
          .await;

        match result {
          Ok(()) => {
            let status = AppUpdateStatus::Downloaded {
              latest_version,
              auto_install_supported: true,
            };
            set_status(app, state, status.clone());
            status
          }
          Err(e) => {
            let status = AppUpdateStatus::Error {
              message: e.to_string(),
            };
            set_status(app, state, status.clone());
            status
          }
        }
      }
      Ok(None) => {
        let status = AppUpdateStatus::Error {
          message: "No in-app update is available for this install.".into(),
        };
        set_status(app, state, status.clone());
        status
      }
      Err(e) => {
        let status = AppUpdateStatus::Error {
          message: e.to_string(),
        };
        set_status(app, state, status.clone());
        status
      }
    },
    Err(e) => {
      let status = AppUpdateStatus::Error {
        message: e.to_string(),
      };
      set_status(app, state, status.clone());
      status
    }
  }
}

pub async fn install_app_update(app: &AppHandle, state: &UpdaterState) -> AppUpdateStatus {
  let last = get_last_status(state);
  if !matches!(
    last,
    AppUpdateStatus::Downloaded {
      auto_install_supported: true,
      ..
    }
  ) {
    return last;
  }

  app.restart();
}
